File: simulation.py
import numpy as np
import matplotlib.pyplot as plt


a = 1664525
c = 1013904223
m = 2**32

def lcg(seed, n=10000):
    x = seed
    nums = []
    for _ in range(n):
        x = (a * x + c) % m
        nums.append(x)
    return np.array(nums)


numbers = lcg(seed=12345, n=10000)


plt.hist(numbers, bins=30, color='pink', edgecolor='black')
plt.title("LCG RNG Histogram")
plt.xlabel("Value")
plt.ylabel("Frequency")
plt.show()


def runs_test(data):
    runs = 1
    for i in range(1, len(data)):
        if (data[i] > data[i - 1]) != (data[i - 1] > data[i - 2] if i > 1 else False):
            runs += 1
    return runs

print("Runs Test:", runs_test(numbers))

File: simulation.rs
use rand::Rng;

const A: u64 = 1664525;
const C: u64 = 1013904223;
const M: u64 = u64::MAX;   // 2^32

fn lcg(seed: u64) -> impl Iterator<Item = u64> {
    let mut x = seed;
    std::iter::from_fn(move || {
        x = (A.wrapping_mul(x).wrapping_add(C)) % M;
        Some(x)
    })
}

fn histogram(nums: &[u64], buckets: usize) -> Vec<usize> {
    let mut hist = vec![0usize; buckets];
    for &n in nums {
        let b = ((n as f64 / M as f64) * buckets as f64) as usize;
        let idx = if b >= buckets { buckets - 1 } else { b };
        hist[idx] += 1;
    }
    hist
}

fn runs_test(nums: &[u64]) -> usize {
    let mut runs = 1;
    for i in 1..nums.len() {
        if nums[i] > nums[i - 1] {
            if nums[i - 1] <= nums.get(i - 2).unwrap_or(&nums[0]) {
                runs += 1;
            }
        }
    }
    runs
}

fn main() {
    // 1. Generate 10,000 numbers
    let numbers: Vec<u64> = lcg(777).take(10_000).collect();

    // 2. Histogram
    let hist = histogram(&numbers, 20);
    println!("Histogram (20 buckets): {:?}", hist);

    // 3. Runs Test
    let r = runs_test(&numbers);
    println!("Runs test result: {r}");

    // Interpretation hints
    println!("High uniformity means histogram buckets are similar size.");
    println!("Runs test should show randomness (not too few or too many runs)");
}